frontend/public/extend/kubevirt/components/vm-list.tsx

```diff
@@ -0,0 +1,166 @@
+  }),
+});
+
+export const menuActions = [menuActionStart, Kebab.factory.Edit, Kebab.factory.Delete];
+
+export const StateColumn = (props) => {
+  const { loaded, loadError, flatten, resources } = props;
+  let value;
+  if (loaded && !loadError) {
+    const vm = flatten(resources);
+    if (vm) {
+      value = getVMStatus(vm);
+    }
+  }
+  return value || <span className="text-muted">{s.notAvailable}</span>;
+};
+
+export const PhaseColumn = (props) => {
+  const { loaded, loadError, flatten, filter, resources } = props;
+  let value;
+  if (loaded && !loadError) {
+    const vmi = filter(flatten(resources));
+    value = _.get(vmi, 'status.phase');
+  }
+  return value || <span className="text-muted">{s.notAvailable}</span>;
+};
+
+export const FirehoseResourceLink = (props) => {
+  const { loaded, loadError, flatten, filter, resources } = props;
+  if (loaded && !loadError) {
+    const data = flatten(resources);
+    if (data) {
+      const resource = filter ? filter(data) : data[0];
+      if (resource) {
+        const { name, namespace } = resource.metadata;
+        const kind = resource.kind || PodModel.kind;
+        // TODO: if `kind` refers to CRD object, use `referenceForModel` function
+        return <ResourceLink kind={kind} name={name} namespace={namespace} />;
+      }
+    }
+  }
+  return <span className="text-muted">{s.notAvailable}</span>;
+};
+
+const VMRow = ({ obj: vm }) => {
+  const vmResource = getFirehoseResource(VirtualMachineModel, {
+    name: vm.metadata.name,
+    namespace: vm.metadata.namespace,
+    isList: false,
+  });
+  const vmiResource = getFirehoseResource(VirtualMachineInstanceModel, {
+    name: vm.metadata.name,
+    namespace: vm.metadata.namespace,
+    isList: true,
+    matchLabels: getLabelMatcher(vm),
+  });
+  const podResource = getFirehoseResource(PodModel, {
+    namespace: vm.metadata.namespace,
+    isList: true,
+    matchLabels: getLabelMatcher(vm),
+  });
+
+  return <ResourceRow obj={vm}>
+    <div className="col-sm-2 col-xs-6">
+      <ResourceLink
+        kind={referenceForModel(VirtualMachineModel)}
+        name={vm.metadata.name}
+        namespace={vm.metadata.namespace}
+        title={vm.metadata.uid} />
+    </div>
+    <div className="col-sm-2 col-xs-6 co-break-word">
+      <ResourceLink
+        kind={NamespaceModel.kind}
+        name={vm.metadata.namespace}
+        title={vm.metadata.namespace} />
+    </div>
+    <div className="col-sm-2 hidden-xs">
+      <Firehose resources={[vmResource]}>
+        <StateColumn flatten={getFlattenForKind(VirtualMachineModel.kind)} />
+      </Firehose>
+    </div>
+    <div className="col-sm-2 hidden-xs">
+      <Firehose resources={[vmiResource]}>
+        <PhaseColumn
+          flatten={getFlattenForKind(VirtualMachineInstanceModel.kind)}
+          filter={data => findVMI(data, vm.metadata.name)} />
+      </Firehose>
+    </div>
+    <div className="col-sm-2 hidden-xs">
+      <Firehose resources={[vmiResource]}>
```

Comment:
FYI, the corresponding code in our fork is [here](https://github.com/kubevirt/web-ui/blob/master/frontend/public/kubevirt/components/vm/vm.jsx).

As I wrote in my [earlier comment](https://github.com/openshift/console/pull/888#issuecomment-453665921), the websocket performance should've been already improved in our fork, but I wanted to avoid pulling in more complexity to get this initial PR in.

@jelkosz I can try to sync with latest code in our fork together with @rawagner.

Reply:
I've worked together with @rawagner on this.

In latest code, VM list doesn't use `Firehose` anymore, and VM details uses `Firehose` only for linking to associated Pod.